mesonbuild/interpreterbase.py

```diff
@@ -228,6 +229,123 @@ def wrapped(*wrapped_args: T.Any, **wrapped_kwargs: T.Any) -> T.Any:
 
+
+def typed_pos_args(name: str, *types: T.Union[T.Type, T.Tuple[T.Type, ...]],
+                   varargs: T.Optional[T.Union[T.Type, T.Tuple[T.Type]]] = None,
+                   optargs: T.Optional[T.List[T.Union[T.Type, T.Tuple[T.Type]]]] = None,
+                   min_varargs: int = 0, max_varargs: int = 0) -> T.Callable[..., T.Any]:
+    """Decorator that types type checking of positional arguments.
+
+    This supports two different models of optional aguments, the first is the
+    variadic argument model. Variadic arguments are a possibly bounded,
+    possibly unbounded number of arguments of the same type (unions are
+    supported). The second is the standard default value model, in this case
+    a number of optional arguments may be provided, but they are still
+    ordered, and they may have different types.
+
+    This function does not support mixing variadic and default arguments.
+
+    :name: The name of the decorated function (as displayed in error messages)
+    :varargs: They type(s) of any variadic arguments the function takes. If
+        None the function takes no variadic args
+    :min_varargs: the minimum number of variadic arguments taken
+    :max_varargs: the maximum number of variadic arguments taken. 0 means unlimited
+    :optargs: The types of any optional arguments parameters taken. If None
+        then no optional paramters are taken.
+
+    allows replacing this:
+    ```python
+    def func(self, node, args, kwargs):
+        if len(args) != 2:
+            raise Exception('... takes exactly 2 arguments)
+        foo: str = args[0]
+        if not isinstance(foo, str):
+            raise ...
+        bar: int = args[1]
+        if not isinstance(bar, int):
+            raise ...
+
+        # actual useful stuff
+    ```
+    with:
+    ```python
+    @typed_pos_args('func_name', str, int)
+    def func(self, node, args: T.Tuple[str, int], kwargs):
+        foo, bar = args
+
+        # actual useful stuff
+    ```
+    """
+    def inner(f: TV_func) -> TV_func:
+
+        @wraps(f)
+        def wrapper(*wrapped_args: T.Any, **wrapped_kwargs: T.Any) -> T.Any:
+            args = _get_callee_args(wrapped_args)[2]
+            assert isinstance(args, list), args
+
+            # These are implementation programming errors, end users should never see them.
+            assert max_varargs >= 0, 'max_varrags cannot be negative'
+            assert optargs is None or varargs is None, \
+                'varargs and opt args not supported together becuase this would be ambigious'
+
+            num_args = len(args)
+            num_types = len(types)
+            a_types = types
+
+            if varargs:
+                num_types += min_varargs
+                if max_varargs == 0 and num_args < num_types:
+                    raise InvalidArguments(f'{name} takes at least {num_types} arguments, but got {num_args}.')
+                elif max_varargs > 0 and num_args > num_types + max_varargs:
```

Comment:
nitpicking: max_varargs is always >0 in this elif case.

I'm not sure `num_args > num_types + max_varargs` is correct because here `num_types` is `len(types) + min_varargs`. If you have between 1 and 2 varargs, it will raise error only if `num_args > 3` instead of `>2`. Unless I'm mistaken. In any case, you have no unit test where you define both max_varargs and min_varargs and you give 1 more arg than the max.

Reply:
hmm, yeah, you're right. now to decide how to fix it :)

Reply:
Actually, max_varargs isn't always >0 in that case, it's still 0 if there are enough arguments, But you're right on the second note, I added tests and they immediately failed :)

Reply:
> Actually, max_varargs isn't always >0 in that case, it's still 0 if there are enough arguments

Right.